content/blog/2020-12-19-LLVM-IR-to-assembly.md

```diff
@@ -0,0 +1,77 @@
+
+# Hardest Part of the Project?
+X86-64 is the hardest part of the project.  Although it is widely adopted, used and the codebase for x86 in LLVM relatively well maintained, the x86-64 ISA is so complex and riddled with special rules and cases that makes any simple modification very painful implement.  It is furthermore complicated by the complexity of the different concurrent calling conventions that x86-64 practices.  Going back to our recent class discussion about rebuild or fix, I think x86 is really something we need consider razing and rebuilding from starch.  It is not x86 is a bad ISA, but the current LLVM codebase in x86 is full of legacy code and code to emulate “bugs” (or features) in GCC compiler.  Although this dates to the history of computer, this makes x86-64 very unfriendly for any new developers to work on it.  Another notable mention is the amount of effort it takes to demystify the target machine files in LLVM.  
+
+# Results and Success?
+The project is a success in that I am able to report the performance compression with both microbenchmarks and on a FFT benchmark with varying number of architectural registers. 
+On the microbenchmarks, I have choosen to compare the number of architectural registers of matrix multiplication in scalar execution, SEE, AVX4 and AVX8 on x86 machine.  For each one of the implementations, they are being compiled by a unmodified version of GCC, unmodified version of clang, and a modified version of clang with only 75% of the original x86-64 registers.    
+The benchmark is primarily done by FFT and is implemented in SSE.   Thie code is compiled by three diffeent compliers: a unmodified version of GCC, unmodified version of clang, and a modified version of clang with only 75% of the original x86-64 registers.     
+The benchmark is done on a 4 core i9-9900K CPU with 16 GB RAM in a Ubuntu VM.  The results are reported cycle count with rdtsc and execution time is reported as wall-clock time.  The results are verified by comparing the results of different implementations.
+
+
+## Microbenchmark 
+
+### Process and Methology
+The microbenchmark preforms a 4x4 matrix multiplication.  A total of 4096 itnerations have been preformed, and the results are aggerate by the minimum function.  Due to the small amount of cycle it takes to compelte the opereation, using wall-clock time may introduced another source of error, therefore I have elected to report the results in CPU cycles.
+
+
+### Results
+Table 1, Microbenchmark of 4x4 matrix in CPU Cycles, 4096 trials, Minimum Number cycles required to complete the operation. 
```

Comment:
How is it possible for a minimum cycle count to be fractional? Maybe this minimum is of an average of several executions?